REGRESSION: source code viewer on Github renders two overlapping carets

rr-codes opened the pull request:
#### 4cc63dfc70a9df30e97dfd5753f8cdad28f96301
<pre>
REGRESSION: source code viewer on Github renders two overlapping carets
<a href="https://bugs.webkit.org/show_bug.cgi?id=259166">https://bugs.webkit.org/show_bug.cgi?id=259166</a>
rdar://112056259

Reviewed by Aditya Keerthi.

The color of the caret should respect both the `caret-color` and the `color` properties.
Because every element implicitly has a `color`, the color of the caret should only reflect
this property if set by the author.

* LayoutTests/fast/css/caret-color-with-color-property-expected.html: Added.
* LayoutTests/fast/css/caret-color-with-color-property.html: Added.
* LayoutTests/fast/css/caret-color-with-inherited-color-property-expected.html: Added.
* LayoutTests/fast/css/caret-color-with-inherited-color-property.html: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::CaretBase::computeCaretColor):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasExplicitlySetColor const):
(WebCore::RenderStyle::InheritedFlags::operator== const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setHasExplicitlySetColor):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueColor):
(WebCore::Style::BuilderCustom::applyInitialColor):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::isAuthorOrigin const):

Canonical link: <a href="https://commits.webkit.org/266070@main">https://commits.webkit.org/266070@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a99653b61ddb18c4f9472e612a0ca72cf91ee963

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12752 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13078 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14491 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12194 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12813 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15579 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13097 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14894 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12916 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13657 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14937 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10945 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11544 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18624 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12020 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11712 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14910 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12171 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10082 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11427 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11414 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3129 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15743 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/12014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/style/StyleBuilderState.h:
```diff
@@ -106,6 +106,7 @@ class BuilderState {
 
     void setIsBuildingKeyframeStyle() { m_isBuildingKeyframeStyle = true; }
 
+    bool isAuthorOrigin() const { return m_currentProperty && m_currentProperty->cascadeLevel == CascadeLevel::Author; };
```
Nit: Add newline.

pxlcoder commented on Source/WebCore/rendering/TextPaintStyle.h:
```diff
@@ -48,6 +48,7 @@ struct TextPaintStyle {
     Color strokeColor;
     Color emphasisMarkColor;
     float strokeWidth { 0 };
+    bool hasExplicitlySetFillColor { false };
```
This is unused.

pxlcoder commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -1589,6 +1589,8 @@ class RenderStyle {
     inline SVGPaintType fillPaintType() const;
     inline StyleColor fillPaintColor() const;
     inline void setFillPaintColor(const StyleColor&);
+    inline void setHasExplicitlySetFillColor(bool);
```
`setHasExplicitlySetColor` / `hasExplicitlySetColor` to match the CSS property name.

mdubet commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -2035,6 +2036,18 @@ inline void BuilderCustom::applyValueColor(BuilderState& builderState, CSSValue&
+    builderState.style().setHasExplicitlySetColor(builderState.isAuthorOrigin());
+}
+
+inline void BuilderCustom::applyInitialColor(BuilderState& builderState)
```
We should have a test case with the CSS `initial` value to test this.
Also, don't we need the same for the `inherit` value ?

mdubet commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -2035,6 +2036,18 @@ inline void BuilderCustom::applyValueColor(BuilderState& builderState, CSSValue&
+    if (builderState.applyPropertyToVisitedLinkStyle())
+        builderState.style().setVisitedLinkColor(RenderStyle::initialColor());
+
+    builderState.style().setDisallowsFastPathInheritance();
```
Why do we disallow fast path inheritance ?

rr-codes replied:
This code was just copy+pasted from the previously-generated code, with the addition of the `isAuthorOrigin` stuff, so I did not change anything to do with this stuff.

mdubet replied:
Thank you for the explanation. Not for this patch obviously, but there might be a good performance gain possible here.

mdubet commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2196,6 +2198,8 @@ class RenderStyle {
         unsigned autosizeStatus : 5;
 #endif
         // 51 bits
+        unsigned hasSetFillColor : 1;
```
Shouldn't this be in the "rare inherited" data structure ? (with text fill and stroke color)

nt1m replied:
The name is misleading, this is not about SVG fill-color, it's about the CSS color property. It's not that uncommon for websites to set the text color afaik, so it shouldn't be in rareData

mdubet replied:
Yes indeed, color is stored `m_inheritedData`.

mdubet replied:
Should we change the name then ?

nt1m replied:
Yes, though I did express disagreement with this general approach on Slack.

whsieh commented on Source/WebCore/style/StyleBuilderState.h:
```diff
@@ -106,6 +106,11 @@ class BuilderState {
+    bool isAuthorOrigin() const
+    {
+        return m_currentProperty && m_currentProperty->cascadeLevel == CascadeLevel::Author;
+    };
```
Nit - stray `;` at the end here.

pxlcoder commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -2034,7 +2036,31 @@ inline void BuilderCustom::applyValueColor(BuilderState& builderState, CSSValue&
         builderState.style().setVisitedLinkColor(resolveColor(color));
     }
+
+    builderState.style().setDisallowsFastPathInheritance();
```
Why is this being added here?

pxlcoder replied:
Oh wait, I think this is just Git showing a weird diff, this is really being added for `applyInheritColor`...

rr-codes replied:
lol ya the diff is quite strangely formatted

pxlcoder commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2196,6 +2198,8 @@ class RenderStyle {
         unsigned autosizeStatus : 5;
 #endif
         // 51 bits
+        unsigned hasSetColor : 1;
```
This should probably be called `hasExplicitlySetColor`.

nt1m commented on LayoutTests/fast/css/caret-color-auto.html:
```diff
@@ -24,7 +24,7 @@
 </head>
 <body>
-<p>Tests that "caret-color: auto" behaves identical to omitting property caret-color.</p>
+<p>Tests that "caret-color: auto" causes the caret to take the "color" property of its parent, if applicable.</p>
```
It's not the color property of the parent, it's the color property of the element itself afaik

nt1m commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -165,6 +165,8 @@ class BuilderCustom {
 
     static void applyValueColor(BuilderState&, CSSValue&);
+    static void applyInitialColor(BuilderState&);
+    static void applyInheritColor(BuilderState&);
```
Can you use the `DECLARE_PROPERTY_CUSTOM_HANDLERS` macro above?

rr-codes committed:
REGRESSION: source code viewer on Github renders two overlapping carets
https://bugs.webkit.org/show_bug.cgi?id=259166
rdar://112056259

Reviewed by Aditya Keerthi.

The color of the caret should respect both the `caret-color` and the `color` properties.
Because every element implicitly has a `color`, the color of the caret should only reflect
this property if set by the author.

* LayoutTests/fast/css/caret-color-with-color-property-expected.html: Added.
* LayoutTests/fast/css/caret-color-with-color-property.html: Added.
* LayoutTests/fast/css/caret-color-with-inherited-color-property-expected.html: Added.
* LayoutTests/fast/css/caret-color-with-inherited-color-property.html: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::CaretBase::computeCaretColor):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasExplicitlySetColor const):
(WebCore::RenderStyle::InheritedFlags::operator== const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setHasExplicitlySetColor):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueColor):
(WebCore::Style::BuilderCustom::applyInitialColor):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::isAuthorOrigin const):

Canonical link: https://commits.webkit.org/266070@main
